apps/bluetooth/js/modules/transfer_manager.js

```diff
@@ -0,0 +1,372 @@
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @param {Object BluetoothAdapter} newAdapter
+     */
+    _onDefaultAdapterChanged: 
+    function bttm__onDefaultAdapterChanged(newAdapter, oldAdapter) {
+      if ((oldAdapter !== null) && (newAdapter === null)) {
+        Debug('_onDefaultAdapterChanged(): Can not get bluetooth adapter!');
+        // Post error with reason then end the activity service.
+        this._endTransferWithReason('cancelled');
+      }
+    },
+
+    /**
+     * 'enabled' change event handler from bluetooth context for
+     * updating UI flow immediately.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @param {Boolean} enabled
+     */
+    _onBluetoothEnabledChanged: 
+    function bttm__onBluetoothEnabledChanged(enabled) {
+      Debug('_onBluetoothEnabledChanged(): enabled = ' + enabled);
+      if (enabled) {
+        // Close the confirmation dialog
+        TurnBluetoothOnDialog.close();
+      } else {
+        TurnBluetoothOnDialog.showConfirm((result) => {
+          if (result === 'confirm') {
+            // Turn Bluetooth on.
+            BtContext.setEnabled(true);
+          } else if (result === 'cancel') {
+            // Post error with reason then end the activity service.
+            this._endTransferWithReason('cancelled');
+          }
+        });
+      }
+    },
+
+    /**
+     * Watch 'devicePicked', 'cancelSelection' event from device picker panel.
+     *
+     * @access private
+     * @memberOf TransferManager
+     */
+    _watchEventFromDevicePicker: function bttm__watchEventFromDevicePicker() {
+      Debug('_watchEventFromDevicePicker():');
+      // Watch 'devicePicked' event for reaching device picked from user.
+      DevicePickerPanel.addEventListener('devicePicked', 
+        this._onDevicePicked.bind(this));
+
+      // Watch 'cancelSelection' event for reaching cancel selection from user.
+      DevicePickerPanel.addEventListener('cancelSelection', 
+        this._onCancelSelection.bind(this));
+    },
+
+    /**
+     * While the device is picked from user, let's send file(s) to the device.
+     *
+     * @access private
+     * @memberOf TransferManager
+     * @param {Object} event
+     * @param {String} event.type - type of event name
+     * @param {Object} event.detail - device info in this object
+     * @param {Object} event.detail.address - address of device
+     */
+    _onDevicePicked: function bttm__onDevicePicked(event) {
+      // TODO: send file to the device via address
+      Debug('_onDevicePicked(): address of target device = ' + 
+            event.detail.address);
+      var targetDeviceAddress = event.detail.address;
+
+      // Produce sending files schedule.
+      // Then, post message to system app.
+      this._produceSendingFilesSchedule();
+
+      // Send each file via Bluetooth sendFile API
+      var blobs = this._activity.source.data.blobs;
+      this._numberOfTasks = blobs.length;
```

Comment:
We can use `Promise.all` to ensure that all of the blobs are transferred, `this._numberOfTasks` can be removed.